docs/scopes.md

```diff
@@ -0,0 +1,199 @@
+* For short-name scope values, split on the ":" character,
+  and A implies B if either:
+  * B[-1] is not "write" and A is a prefix of B, or.
+  * A[-1] is "write", and:
+    * A[:-1] is a prefix of B, or
+    * B[-1] is "write" and A[:-1] is a prefix of B[:-1]
+
+More precisely, the algoritm for checking implication is:
+
+* If A is a `https://` URL, then:
+  * If B is not a `https://` URL, then fail.
+  * If the origin of B is different than that of A, then fail.
+  * If the path component list of A is not a prefix of the path
+    component list of B, then fail.
+  * If A has a fragment, then:
+    * If B does not have a fragment, then fail.
+    * If B has a fragment that differs from A, then fail.
+  * Otherwise, succeed.
+* Otherwise:
+  * If B is a `https://` URL, then fail.
+  * Split A and B into components based on `:` delimiter.
+  * If the last component of B is `write`, then:
+    * If the last component of A is not `write`, then fail.
+  * If the last component of A is `write`, remove it.
+  * If A is not a prefix of B, then fail.
+  * Otherwise, succeed.
+
+Below are some testcases against which
+scope-checking code can be validated.
+
+Valid implications:
+* `profile:write` implies `profile`.
+* `profile` implies `profile:email`.
+* `profile:write` implies `profile:email`.
+* `profile:write` implies `profile:email:write`.
+* `profile:email:write` implies `profile:email`.
+* `profile profile:email:write` implies `profile:email`.
+* `profile profile:email:write` implies `profile:display_name`.
+* `profile https://identity.mozilla.com/apps/oldsync` implies `profile`.
+* `profile https://identity.mozilla.com/apps/oldsync` implies `https://identity.mozilla.com/apps/oldsync`.
+* `https://identity.mozilla.com/apps/oldsync` implies `https://identity.mozilla.com/apps/oldsync#read`.
+* `https://identity.mozilla.com/apps/oldsync` implies `https://identity.mozilla.com/apps/oldsync/bookmarks`.
+* `https://identity.mozilla.com/apps/oldsync` implies `https://identity.mozilla.com/apps/oldsync/bookmarks#read`.
+* `https://identity.mozilla.com/apps/oldsync#read` implies `https://identity.mozilla.com/apps/oldsync/bookmarks#read`.
+* `https://identity.mozilla.com/apps/oldsync#read profile` implies `https://identity.mozilla.com/apps/oldsync/bookmarks#read`.
+
+Invalid implications:
+* `profile:email:write` does *not* imply `profile`.
+* `profile:email:write` does *not* imply `profile:write`.
+* `profile:email` does *not* imply `profile:display_name`.
+* `profilebogey` does *not* imply `profile`.
+* `profile:write` does *not* imply `https://identity.mozilla.com/apps/oldsync`.
+* `profile profile:email:write` does *not* imply `profile:write`.
+* `https` does *not* imply `https://identity.mozilla.com/apps/oldsync`.
+* `https://identity.mozilla.com/apps/oldsync` does *not* imply `profile`.
+* `https://identity.mozilla.com/apps/oldsync#read` does *not* imply `https://identity.mozila.com/apps/oldsync/bookmarks`.
```

Comment:
In the URL after "does not imply" do you want "mozila" or "mozilla"? Applies to this and the following 6 lines.

Reply:
should be `mozilla ` yeap :)